Sharing - add link to customize share button settings

blackjackkent opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

Fixes 116-gh-dotcom-manage

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* This PR adds a link under the "Share This" buttons on post pages (for users with sufficient permissions) to the customization screen for those buttons, to make it easier for users to realize that they are indeed customizable.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
N/A

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Enable share buttons on Jetpack-enabled site by navigating to WP Admin -> Jetpack -> Settings -> Sharing and enabling the "Add sharing buttons to your posts and pages" toggle.
* Navigate to a post on the site, both while logged in as a user with admin permissions (`manage_options`, specifically) and while not.
* Verify that while logged in with permissions, a "Customize buttons" link is displayed under the sharing buttons which links the user to the customization options in Calypso.
* Verify that this option does not exist if the user does not have admin permissions.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Added direct link to share button customization for logged-in admins.

blackjackkent commented:
This is ready for testing in Jetpack. 

@Automattic/jetpack-crew Do you have any insight on how to fix the Fusion issue on `modules/sharedaddy/sharing.css`? The diff looks clean but it's claiming it can't sync.

gwwar commented:
@blackjackkent is this one ready for review? We can probably turn this into a PR vs draft if so and I can test this tomorrow.

kwight commented on modules/sharedaddy/sharing-service.php:
```diff
@@ -867,6 +867,13 @@ function sharing_display( $text = '', $echo = false ) {
 
+			// Link to customization options if user can manage them.
+			if ( current_user_can( 'manage_options' ) ) {
+				$link_text        = esc_html__( 'Customize buttons', 'jetpack' );
```
With no dynamic elements to the string literal, we can just go with `__()`.

In WordPress, it's considered best practice to escape as late as possible, so in this case, we can leave `$link_text` and `$link_url` without any escaping for simplicity, and then escape when concatenating to `$sharing_content` (which also covers any later changes to those variables without worry).

VIP has some great docs around escaping: https://wpvip.com/documentation/vip-go/validating-sanitizing-and-escaping/#always-escape-late

kwight commented on modules/sharedaddy/sharing-service.php:
```diff
@@ -867,6 +867,13 @@ function sharing_display( $text = '', $echo = false ) {
+			if ( current_user_can( 'manage_options' ) ) {
+				$link_text        = esc_html__( 'Customize buttons', 'jetpack' );
+				$link_url         = 'https://jetpack.com/redirect/?source=calypso-marketing-sharing-buttons&site=' . Jetpack::build_raw_urls( get_home_url() );
+				$sharing_content .= '<p class="share-customize-link"><a href="' . $link_url . '" target="_blank" rel="noopener noreferrer">' . $link_text . '</a></p>';
```
So this would have `esc_url( $link_url )` and `esc_html( $link_text )` for good measure.

blackjackkent commented on modules/sharedaddy/sharing.css:
```diff
@@ -753,3 +753,14 @@ div.sharedaddy.sharedaddy-dark #sharing_email {
+}
+
+.sd-content .share-customize-link a{
+	font-size: 11px !important;
```
Yeah, I tend to avoid it as well; in this case the issue requirements mentioned matching the style used on the "{x} bloggers liked this" text on the Likes widget; I copied these styles from there, where we were also using `!important`, so I assumed the intention is to make sure that we are overriding any theme link styles. Not sure though.

gwwar replied:
Looks like we're adding this markup? Overall, it feels okay either way, if other themes were building on top of this class, we'd probably require the `!important`

gwwar replied:
🤔 overall I'm not sure how much we care about the font size sticking to 11px vs scaling with the theme. For example, notice how large the Edit control is here:

<img width="661" alt="Screen Shot 2020-07-15 at 10 25 46 AM" src="https://user-images.githubusercontent.com/1270189/87576080-b3efab00-c685-11ea-8249-648e82ab9049.png">

kwight replied:
Since the spirit of the styling comes from the "(x bloggers) like this" text below Likes, I think pinning it is reasonable (that Like text is pinned to the same values).

kwight commented:
Works well, we're close!

I noticed some styling oddness; it may be best to drop the hardcoded black color, since it seems it should inherit from the body. And the anchor color for the "likes this" might be determined server-side (I didn't dig too far, let me know if you need a hand seeing where that's all coming from). I got this display configuration by having two other test users logged in and liking the post (this is what the admin sees).

<img width="1374" alt="Screen Shot 2020-07-15 at 10 05 33 AM" src="https://user-images.githubusercontent.com/349751/87574598-72f69700-c683-11ea-8419-0a917e558136.png">

blackjackkent commented:
Updated per comments above. I removed the `!important` and font-color stylings; end result looks like this on the default theme:

<img width="303" alt="Screen Shot 2020-07-15 at 2 43 50 PM" src="https://user-images.githubusercontent.com/13437011/87588668-a9dfa380-c6a9-11ea-9e87-bbbbf9ab240a.png">

blackjackkent commented:
Corresponding diff for wpcom opened at D46447-code .

kwight commented on modules/sharedaddy/sharing.css:
```diff
@@ -753,3 +753,13 @@ div.sharedaddy.sharedaddy-dark #sharing_email {
+	line-height: 11px;
+}
+
+.sd-content .share-customize-link a{
```
Nit: missing space between `a{`.

blackjackkent replied:
Fixed!

kwight commented:
Sweet!

Let's go with this 👍  There are some differences in behaviour with the Likes below (assuming they're active), but that's because this new link falls under purview of the theme's stylesheet, while the Likes text does some server-side trickery to approximate theme-compatible colours (it has no choice, being served in an iframe). This can lead to different hover behaviour in particular, but because the Likes mechanism isn't up to par.

![2020-07-16 13 09 25](https://user-images.githubusercontent.com/349751/87717841-eb358900-c765-11ea-85dd-e020f880f762.gif)

kwight commented:
D46447-code looks problematic on the dotcom side, which will likely need an update here before merging.

blackjackkent commented:
@kwight Looking into the phabricator issue now!

jeherve commented on modules/sharedaddy/sharing-service.php:
```diff
@@ -867,6 +867,13 @@ function sharing_display( $text = '', $echo = false ) {
+			// Link to customization options if user can manage them.
+			if ( current_user_can( 'manage_options' ) ) {
+				$link_text        = __( 'Customize buttons', 'jetpack' );
+				$link_url         = 'https://jetpack.com/redirect/?source=calypso-marketing-sharing-buttons&site=' . Jetpack::build_raw_urls( get_home_url() );
```
I would recommend using `\Automattic\Jetpack\Redirect::get_url()` to build that URL.

See PCYsg-pY7-p2 for more info.

I also wonder if we should link to the sharing settings in wp-admin for Jetpack site owners (non-Atomic)?

jeherve commented:
> Do you have any insight on how to fix the Fusion issue on modules/sharedaddy/sharing.css? The diff looks clean but it's claiming it can't sync.

Looks like they're a bit out of sync, but D46551-code should do the trick now.

pablinos commented:
I've updated this to use the `Redirect` class. @jeherve mentioned sending non-Atomic Jetpack users to wp-admin, but the customise buttons link in the settings of my Jetpack site takes me to Calypso too, so if I'm understanding it correctly, I think we're fine to always send people to Calypso.

gwwar commented on modules/sharedaddy/sharing.css:
```diff
@@ -753,3 +753,13 @@ div.sharedaddy.sharedaddy-dark #sharing_email {
 }
+
+.sd-content .share-customize-link {
+	margin-top: -.7em;
```
Let's not use negative margin here, depending on the theme this is going to overlap with the buttons.

<img width="640" alt="Screen Shot 2020-07-17 at 1 56 23 PM" src="https://user-images.githubusercontent.com/1270189/87830210-729dfd80-c835-11ea-96b7-3393917eab05.png">

gwwar replied:
Maybe `margin: 0.1em`? Though, I'll defer to @sixhours or @danhauk here.

<img width="504" alt="Screen Shot 2020-07-17 at 2 01 15 PM" src="https://user-images.githubusercontent.com/1270189/87830500-01127f00-c836-11ea-91f9-ba4cb7e43ad4.png">

gwwar commented:
Thanks for creating the fusion patch @jeherve and @pablinos for updating the URL. 

I re-tested this and D46551-code and verified that on Simple/Jetpack we navigate to the right settings screen in Calypso. @blackjackkent We just need to fix the margin styling, and I think we should be all set.

kraftbj commented:
r210720-wpcom

eurello commented:
I'm excited about this change, I think it'll be really helpful to customers - thanks for doing it!
